[Model Element] Require CORS for loading model sources and environment maps

achan00 opened the pull request:
#### 74ab595af1f056614e05dd61124a38c7288ca5a3
<pre>
[Model Element] Require CORS for loading model sources and environment maps
<a href="https://bugs.webkit.org/show_bug.cgi?id=293063">https://bugs.webkit.org/show_bug.cgi?id=293063</a>
<a href="https://rdar.apple.com/151364606">rdar://151364606</a>

Reviewed by Youenn Fablet and Anne van Kesteren.

Make sure CORS is always enabled with model source and environment
map requests, by passing a non-null crossorigin attribute to
createPotentialAccessControlRequest().

Also make sure we null out m_model once source is reset.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-allowed-with-all-access-header-expected.txt: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-allowed-with-all-access-header.html: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-allowed-with-header-expected.txt: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-allowed-with-header.html: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-allowed-with-same-origin-expected.txt: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-allowed-with-same-origin.html: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-blocked-with-header-expected.txt: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-blocked-with-header.html: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-blocked-with-use-credentials-and-all-access-header-expected.txt: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-blocked-with-use-credentials-and-all-access-header.html: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-blocked-without-header-expected.txt: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-blocked-without-header.html: Added.
* LayoutTests/http/tests/security/resources/cube.usdz: Added.
* LayoutTests/http/tests/security/resources/model-access-control.py: Added.
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::setSourceURL):
(WebCore::HTMLModelElement::environmentMapRequestResource):

Canonical link: <a href="https://commits.webkit.org/295219@main">https://commits.webkit.org/295219@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2409c48413b3c530ecc2f61b94223ac3648676de

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104460 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24169 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14587 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109667 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/55131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106500 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24551 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32714 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/79320 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/55131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107466 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19094 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94282 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59646 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12354 "Passed tests") | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54498 "Failed to compile WebKit") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12412 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112052 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31621 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/23294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/88361 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31985 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90515 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88026 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22411 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32919 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10696 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/26096 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31548 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/36901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31340 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34679 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/32900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

annevk reviewed (approved):
This looks good to me, but let's wait for Youenn as well to be sure.

youennf reviewed (approved):
LGTM with a test where load would fail.

youennf commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -187,6 +188,22 @@ void HTMLModelElement::sourcesChanged()
     setSourceURL(selectModelSource());
 }
 
+CachedResourceRequest HTMLModelElement::createResourceRequest(const URL& resourceURL, const FetchOptions::Destination& destination)
```
```suggestion
CachedResourceRequest HTMLModelElement::createResourceRequest(const URL& resourceURL, FetchOptions::Destination destination)
```

achan00 replied:
I'll add more tests in a separate PR.

achan00 replied:
Filed: https://bugs.webkit.org/show_bug.cgi?id=293371 (Add more tests around model source loading across different origins)

youennf reviewed (approved):
LGTM.
If you have some time, we could add a few additional tests:
- a data URL load from a HTTP origin (should succeed)
- a blob URL load from a HTTP origin (should succeed)
- a file URL load from a file origin (should succeed)
- a file URL load with a crossorigin attribute (whatever value, should fail)
- a HTTP URL load with an empty crossorigin attribute (should be the same as anonymous).

achan00 committed:
[Model Element] Require CORS for loading model sources and environment maps
https://bugs.webkit.org/show_bug.cgi?id=293063
rdar://151364606

Reviewed by Youenn Fablet and Anne van Kesteren.

Make sure CORS is always enabled with model source and environment
map requests, by passing a non-null crossorigin attribute to
createPotentialAccessControlRequest().

Also make sure we null out m_model once source is reset.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-allowed-with-all-access-header-expected.txt: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-allowed-with-all-access-header.html: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-allowed-with-header-expected.txt: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-allowed-with-header.html: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-allowed-with-same-origin-expected.txt: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-allowed-with-same-origin.html: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-blocked-with-header-expected.txt: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-blocked-with-header.html: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-blocked-with-use-credentials-and-all-access-header-expected.txt: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-blocked-with-use-credentials-and-all-access-header.html: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-blocked-without-header-expected.txt: Added.
* LayoutTests/http/tests/security/model-element/model-element-crossorigin-blocked-without-header.html: Added.
* LayoutTests/http/tests/security/resources/cube.usdz: Added.
* LayoutTests/http/tests/security/resources/model-access-control.py: Added.
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::setSourceURL):
(WebCore::HTMLModelElement::environmentMapRequestResource):

Canonical link: https://commits.webkit.org/295219@main
